pkg/server/commands/reverseexpand/reverse_expand.go

```diff
@@ -421,7 +422,13 @@ func (c *ReverseExpandQuery) readTuplesAndExecute(
 	if err != nil {
 		return err
 	}
-	defer iter.Stop()
+
+	// filter out invalid tuples yielded by the database iterator
+	filteredIter := storage.NewFilteredTupleKeyIterator(
+		storage.NewTupleKeyIteratorFromTupleIterator(iter),
+		validation.FilterInvalidTuples(c.typesystem),
```

Comment:
I think that, for safety and for speed, we should only remove tuples that are invalid because of the condition.

```go
func FilterInvalidConditionedTuples(typesys *typesystem.TypeSystem) storage.TupleKeyFilterFunc {
	return func(tupleKey *openfgav1.TupleKey) bool {
		err := ValidateCondition(typesys, tupleKey)
		return err == nil
	}
}
```

Also, in the future i think it's worth revisiting `ReadStartingWithUser` to include another filter by `condition`. This way we avoid the unnecessary DB roundtrips in the first place.

Reply:
>  in the future i think it's worth revisiting ReadStartingWithUser to include another filter by condition. This way we avoid the unnecessary DB roundtrips in the first place.

Agreed on this point.

> I think that, for safety and for speed, we should only remove tuples that are invalid because of the condition.

Fair point - my goal was to try and keep it more consistent with already tested behavior we depend on in Check, but a quick and dirty check directly on the defined condition is just as good. Will updated.